compiler-plugin/src/main/java/io/ballerina/lib/data/csvdata/compiler/CsvDataTypeValidator.java

```diff
@@ -0,0 +1,649 @@
+        if (mappingConstructorExprNodeCount > 1) {
+            assert false : "MappingConstructorExpressionNode count in the function " +
+                    "arguments should be less than or equal to 1";
+        }
+    }
+
+    private void throwErrorsIfIgnoredFieldFoundForOutputs(String header, String customHeadersIfHeaderAbsent,
+            String headerRows, String customHeaders, boolean isCustomHeaderPresent, String headerOrder,
+            String outputWithHeaders, SyntaxNodeAnalysisContext ctx, Location currentLocation,
+            String functionName, boolean isRecord) {
+        if (functionName.equals(Constants.PARSE_STRING) && header != null && !header.equals(Constants.FALSE)
+                && customHeadersIfHeaderAbsent != null && !customHeadersIfHeaderAbsent.equals(Constants.NIL)
+                && !customHeadersIfHeaderAbsent.equals(Constants.NULL)) {
+            reportDiagnosticInfo(ctx, Optional.ofNullable(currentLocation),
+                    CsvDataDiagnosticCodes.IGNORE_CUSTOM_HEADERS_PARAMETER_WHEN_HEADER_PRESENT);
+        }
+        if (functionName.equals(Constants.PARSE_LISTS) && headerRows != null
+                && !headerRows.equals("0") && !headerRows.equals("1") &&
+                (!isCustomHeaderPresent || (customHeaders != null && (customHeaders.equals(Constants.NIL)
+                || customHeaders.equals(Constants.NULL))))) {
+            reportDiagnosticInfo(ctx, Optional.ofNullable(currentLocation),
+                    CsvDataDiagnosticCodes.CUSTOM_HEADERS_SHOULD_BE_PROVIDED);
+        }
+        if (isRecord && functionName.equals(Constants.TRANSFORM) && headerOrder != null
+                && !headerOrder.equals(Constants.NIL) && !headerOrder.equals(Constants.NULL)) {
+            reportDiagnosticInfo(ctx, Optional.ofNullable(currentLocation),
+                    CsvDataDiagnosticCodes.IGNORE_HEADERS_ORDER_FOR_RECORD_ARRAY);
+        }
+        if (isRecord && outputWithHeaders != null) {
+            reportDiagnosticInfo(ctx, Optional.ofNullable(currentLocation),
+                    CsvDataDiagnosticCodes.IGNORE_OUTPUT_HEADERS_FOR_RECORD_ARRAY);
+        }
+    }
+
+    private String getTheValueOfTheUserConfigOption(SpecificFieldNode specificFieldNode) {
+        Optional<ExpressionNode> optExpNode = specificFieldNode.valueExpr();
+        if (optExpNode.isPresent()) {
+            ExpressionNode expNode = optExpNode.get();
+            if (expNode instanceof BasicLiteralNode basicLiteralNode) {
+                return basicLiteralNode.literalToken().text();
+            }
+            if (expNode instanceof ListConstructorExpressionNode listConstructorExpressionNode) {
+                return listConstructorExpressionNode.expressions().toString();
+            }
+            if (expNode instanceof NilLiteralNode) {
+                return Constants.NIL;
+            }
+        }
+        return null;
```

Comment:
I think you can use the monadic form here as well. Also when refactoring `validateFunctionParameterTypesWithOptions` shall we see if it is better for this to return Optional instead of string. Mixing Optinals and nullable values in the same class looks a bit incosistent IMO.

```suggestion
        return specificFieldNode.valueExpr().map(expNode -> {
            if (expNode instanceof BasicLiteralNode basicLiteralNode) {
                return basicLiteralNode.literalToken().text();
            }
            if (expNode instanceof ListConstructorExpressionNode listConstructorExpressionNode) {
                return listConstructorExpressionNode.expressions().toString();
            }
            if (expNode instanceof NilLiteralNode) {
                return Constants.NIL;
            }
            return null;
        }).orElse(null);
```